src/mlpack/core/math/shuffle_data.hpp

```diff
@@ -67,11 +70,80 @@ void ShuffleData(const MatType& inputPoints,
     ++index;
   }
 
-  outputPoints = MatType(locations, values, inputPoints.n_rows,
-      inputPoints.n_cols, true);
-  outputLabels.set_size(inputLabels.n_elem);
-  for (size_t i = 0; i < inputLabels.n_elem; ++i)
-    outputLabels[ordering[i]] = inputLabels[i];
+  if (&inputPoints == &outputPoints || &inputLabels == &outputLabels)
+  {
+    MatType newOutputPoints(locations, values, inputPoints.n_rows,
+        inputPoints.n_cols, true);
+    LabelsType newOutputLabels(inputLabels.n_elem);
+    for (size_t i = 0; i < inputLabels.n_elem; ++i)
+      newOutputLabels[ordering[i]] = inputLabels[i];
```

Comment:
Can you use a non-contiguous view here, to avoid the for loop? It's probably the same.

Reply:
Ah, I didn't think of that---fixed.